app/src/main/java/com/github/sdp/ratemyepfl/database/reviewable/EventRepositoryImpl.kt

```diff
@@ -0,0 +1,109 @@
+
+    @Inject
+    constructor(db: FirebaseFirestore) : this(
+        ReviewableRepositoryImpl(
+            db,
+            EVENT_COLLECTION_PATH,
+            NAME_FIELD_NAME,
+        ) { documentSnapshot ->
+            documentSnapshot.toEvent()
+        })
+
+
+    companion object {
+        const val NAME_FIELD_NAME: String = "name"
+        const val EVENT_COLLECTION_PATH = "events"
+        const val LATITUDE_FIELD_NAME = "lat"
+        const val LONGITUDE_FIELD_NAME = "long"
+        const val NUMBER_PARTICIPANTS_FIELD_NAME = "numParticipants"
+        const val LIMIT_PARTICIPANTS_FIELD_NAME = "limitParticipants"
+        const val DATE_FIELD_NAME = "date"
+
+        fun DocumentSnapshot.toEvent(): Event? {
+            val name = getString(NAME_FIELD_NAME)
+            val lat = getDouble(LATITUDE_FIELD_NAME)
+            val long = getDouble(LONGITUDE_FIELD_NAME)
+            val numReviews = getField<Int>(NUM_REVIEWS_FIELD_NAME)
+            val averageGrade = getDouble(AVERAGE_GRADE_FIELD_NAME)
+            val numParticipants = getField<Int>(NUMBER_PARTICIPANTS_FIELD_NAME)
+            val limitParticipants = getField<Int>(LIMIT_PARTICIPANTS_FIELD_NAME)
+            val date = LocalDateTime.parse(getString(DATE_FIELD_NAME))
+            return try {
+                Event.Builder(
+                    name,
+                    numReviews,
+                    averageGrade,
+                    numParticipants,
+                    limitParticipants,
+                    lat,
+                    long,
+                    date
+                )
+                    .build()
+            } catch (e: IllegalStateException) {
+                null
+            }
+        }
+    }
+
+    override suspend fun getEvents(): List<Event> = take(DEFAULT_QUERY_LIMIT.toLong())
+        .mapNotNull { it.toEvent() }
+
+    override suspend fun getEventById(id: String): Event? = getById(id).toEvent()
+
+    override suspend fun incrementParticipants(id: String) {
+        coroutineScope {
+            this.launch(Dispatchers.IO) {
+                changeParticipants(id, 1).await()
+            }
+        }
+    }
+
+    override suspend fun decrementParticipants(id: String) {
+        coroutineScope {
+            this.launch(Dispatchers.IO) {
+                changeParticipants(id, -1).await()
+            }
+        }
+    }
+
+    override suspend fun updateEventRating(id: String, rating: ReviewRating) {
+        repository.update(id) { event ->
+            val (updatedNumReview, updatedAverageGrade) = ReviewableRepositoryImpl.computeUpdatedRating(
+                event,
+                rating
+            )
+            event.copy(numReviews = updatedNumReview, averageGrade = updatedAverageGrade)
+        }.await()
+
+    }
+
+    private fun changeParticipants(id: String, incDec: Int): Task<Transaction> =
+        repository.update(id) { event ->
+            event.copy(numParticipants = event.numParticipants + incDec)
+        }

```

Comment:
Couldn't you have left the "await()" here instead of putting it twice above?

Reply:
Here it returns a Task<Transaction>, and I wanted to expose the asynchrony. I think we may change that later (typically force increment/decrement) to be called asynchronousy.

Reply:
Yeah for sure, I will make them asynchronous and use callbacks in the view model to update the events.